opto/trainer/loader.py

```diff
@@ -24,26 +24,40 @@ def __init__(self, dataset, batch_size=1, randomize=True, replacement=False, shu
         self.randomize = randomize
         self.replacement = replacement
         self.shuffle = shuffle
-        self._indices = self._update_indices()
-        self.n_epochs = 0
+        self.n_epochs = -1
         self._i = 0
+        self._indices = [ i for i in range(len(self.dataset['inputs'])) ]
+        self._exhausted = False
+        self._start_new_epoch()  # self.n_epochs will be set to 0
+
+    def _start_new_epoch(self):
+        if self.shuffle:
+            self._indices = self._update_indices()
+        self._i = 0
+        self.n_epochs += 1
+        self._exhausted = False
 
     def __iter__(self):
         return self
 
     def __next__(self):
-        """ Get the next batch of data """
-        if self._i >= len(self._indices):
-            if self.shuffle:
-                self._indices = self._update_indices()
-            # Reset the index for the next epoch
-            self._i = 0
-            self.n_epochs += 1
+        """Get the next batch of data, always of batch_size. If the dataset is smaller or at the end, the batch will include data from the next epoch after shuffling."""
```

Comment:
Wait, is this the logic we want? 
This will cause the `step` vs `epoch` distinction. I think a lot of trainer does this...so I'm ok if this is the route we want to go with...but just curious if there's any special rationale for this

Reply:
I see -- that makes sense!